### PR TITLE
Enhance default provider documentation

### LINKED ISSUE #13844: OSSL_PROVIDER_do_all creating side effects
This pertains to a locally build (linux-x86_64) master (OpenSSL3): The return value of [`OSSL_PROVIDER_available`](https://www.openssl.org/docs/manmaster/man3/OSSL_PROVIDER_available.html) depends on whether (apparently side-effect free `OSSL_PROVIDER_do_all`) has been run before (or not). This is at least unexpected; if there is an explanation, I'd be glad to get it.

Code to reproduce:
```
#include <string.h>

#include <openssl/crypto.h>
#include <openssl/provider.h>

int display_provider(OSSL_PROVIDER *provider, void *cbdata) {
   printf("Provider loaded: %s\n", OSSL_PROVIDER_name(provider));
   return 1;
}

int main(int argc, char *argv[]) {
    OSSL_LIB_CTX *libctx = OSSL_LIB_CTX_new();

    if (!libctx)
        return -1;

    // run this iterator only if two parameters are passed:
    if (argc>2) OSSL_PROVIDER_do_all(libctx, display_provider, NULL);

    if (!OSSL_LIB_CTX_load_config(libctx, "base.cnf"))
        return -1;

    // run this iterator if any parameter is passed:
    if (argc>1) OSSL_PROVIDER_do_all(libctx, display_provider, NULL);

    printf("default provider available: %d\n", 
            OSSL_PROVIDER_available(libctx, "default"));

}

```

Compiled with `gcc -Iinclude bug.c -L. -lcrypto`, this command sequence shows the problem:
```
> LD_LIBRARY_PATH=. ./a.out 
default provider available: 0
> LD_LIBRARY_PATH=. ./a.out 1
Provider loaded: base
default provider available: 0
> LD_LIBRARY_PATH=. ./a.out 1 2
Provider loaded: default
Provider loaded: base
Provider loaded: default
default provider available: 1

```
The first invocation returns as expected: "default" provider is not loaded (as per cnf below).
The second invocation now iterates providers after loading the configuration and only the loaded "base" is shown; "default" is correctly shown as absent.
The final iteration already at the start shows "default" loaded when iterating providers (and subsequently returns a different result in `OSSL_PROVIDER_available`). This does not seem plausible and is not explicable from [the documentation](https://www.openssl.org/docs/manmaster/man3/OSSL_PROVIDER_do_all.html): Or in other words: Why does calling `OSSL_PROVIDER_do_all` cause the "default" provider to be loaded?

`base.cnf` contents:
```
openssl_conf = openssl_init

[openssl_init]
providers = provider_sect

[provider_sect]
base = base_sect

[base_sect]
activate = 1
```

### PR DESCRIPTION
**Pull request by baentsch**
Bring Wiki and man page documentation in line regarding default provider fall-back behaviour.
Also added return value description of `OSSL_PROVIDER_do_all`.

Fixes #13844

- [x] documentation is added or updated

### CODE REVIEW

File: doc/man3/OSSL_PROVIDER.pod
Comment (mattcaswell) [inline doc/man3/OSSL_PROVIDER.pod:97]: I think we've tended to use the word "fallback", i.e. without a hyphen

File: doc/man7/OSSL_PROVIDER-default.pod
Comment (mattcaswell) [inline doc/man7/OSSL_PROVIDER-default.pod:12]: fallback

File: doc/man3/OSSL_PROVIDER.pod
Comment (mattcaswell) [inline doc/man3/OSSL_PROVIDER.pod:149]: I'm not sure how to read "if a first callback". Do you mean "if any callback"?

File: doc/man7/OSSL_PROVIDER-default.pod
Comment (mattcaswell) [inline doc/man7/OSSL_PROVIDER-default.pod:13]: requested from a provider -> fetched

File: doc/man7/OSSL_PROVIDER-default.pod
Comment (mattcaswell) [inline doc/man7/OSSL_PROVIDER-default.pod:14]: There is some subtlety here.

We only attempt to load the fallback once. So if it is subsequently unloaded and no providers are available, we don't try again.

Also any attempt to explicitly load a provider (either via API or config) will disable subsequent fallback loading, even if the load attempt fails.

**Review by mattcaswell — CHANGES_REQUESTED** [review]
I think some reference to the OSSL_PROVIDER-default.pod text should be made in the "Fetching algorithms" section of provider.pod. Basically saying that if any algorithm is explicitly or implicitly fetched then the fallback provider may be automatically loaded.

File: doc/man3/OSSL_PROVIDER.pod
Comment (baentsch) [reply doc/man3/OSSL_PROVIDER.pod:149]: Well, that's the thing: The `do_all` code is calling the callback function in sequence for each provider found and stops doing so if one (the first) callback returns 0. So I considered the word "any" in the second part to be misleading as it somehow implies the callback would be called for all providers. But then again, I am no native English speaker :-]

File: doc/man3/OSSL_PROVIDER.pod
Comment (baentsch) [reply doc/man3/OSSL_PROVIDER.pod:97]: Well, only in man7: There it's "fallback" throughout; in man3 it's been mostly "fall-back". But I also like "fallback" better. OK to change the other "fall-back" instances too?

File: doc/man7/OSSL_PROVIDER-default.pod
Comment (baentsch) [reply doc/man7/OSSL_PROVIDER-default.pod:12]: ACK

File: doc/man7/OSSL_PROVIDER-default.pod
Comment (baentsch) [reply doc/man7/OSSL_PROVIDER-default.pod:13]: Done

File: doc/man7/OSSL_PROVIDER-default.pod
Comment (baentsch) [reply doc/man7/OSSL_PROVIDER-default.pod:14]: OK - upon reading the Wiki, quite a lot of subtleties here... I'll try to amend....

Comment (baentsch) [reply doc/man7/OSSL_PROVIDER-default.pod:14]: Done. Please check again.

File: doc/man7/OSSL_PROVIDER-default.pod
Comment (mattcaswell) [inline doc/man7/OSSL_PROVIDER-default.pod:21]: These are slightly separate points: i.e. automatic loading occurs only once, and it won't happen if a load attempt has already been made. The automatic loading is a maximum of once. If a load attempt has been made then it won't happen at all. How about this slight reformulation:

> If an attempt to load a provider has already been made (whether successful or not)  then the default provider won't be loaded automatically. Therefore if it is to be used in conjunction with other providers then it must be loaded explicitly. Automatic loading of the default provider only occurs a maximum of once; if the default provider is explicitly unloaded then the default provider will not be automatically loaded again.

File: doc/man3/OSSL_PROVIDER.pod
Comment (mattcaswell) [inline doc/man3/OSSL_PROVIDER.pod:150]: I know now what you're trying to say, but I think it doesn't quite sound right. How about:

"callback processing stops at the first callback invocation on a provider that returns 0"

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: doc/man3/OSSL_PROVIDER.pod
Comment (richsalz) [reply doc/man3/OSSL_PROVIDER.pod]: If you change fall-back to fallback, consider a PR that adds this to the `preferred_words` list in `util/find-doc-nits`

File: doc/man3/OSSL_PROVIDER.pod
Comment (baentsch) [reply doc/man3/OSSL_PROVIDER.pod]: Thanks. So inserted.

File: doc/man7/OSSL_PROVIDER-default.pod
Comment (baentsch) [reply doc/man7/OSSL_PROVIDER-default.pod]: Much better. Thanks! Sorry for the iterations.

File: doc/man3/OSSL_PROVIDER.pod
Comment (baentsch) [reply doc/man3/OSSL_PROVIDER.pod]: Didn't know this existed: Thanks for pointing to it. So done.

**Review by mattcaswell — APPROVED** [review]
LGTM

**Comment by mattcaswell** [comment]
Needs a second review!

**Review by t8m — APPROVED** [review]

**Comment by mattcaswell** [comment]
Pushed! Thanks for the contribution.

**Pull request closed without merging**
